IPAM: handle subnets bigger than "/64"

robmry opened the pull request:
**- What I did**

- fix https://github.com/moby/moby/issues/21199
- fix https://github.com/moby/moby/issues/35195
- fix https://github.com/moby/moby/issues/45402
- fix https://github.com/moby/moby/issues/48527
- replaces the workaround added in https://github.com/moby/moby/pull/48322

The default IPAM allocator is unable to represent subnets larger than 64-bits (subnets with a smaller prefix), because it uses a Bitmap that's limited to 64-bits.

When it's used to represent a 64-bit subnet, the top address can't be allocated (because bitmap.Bitmap is initialised with the number of bits it needs to represent in a uint64, so it's one short).

The rest of the daemon doesn't know about these limitations, so strange things happen when a large IPv6 subnet is used. No errors are reported, addresses/subnets are just set up incorrectly.  The IPAM code calculates offsets into the bitmap itself, details it shouldn't need to understand and, because it's working on offsets into a range it doesn't always notice when it's asked to set a bit outside the range (see the notes about "offset" in https://github.com/moby/moby/pull/48319).

It's unusual to need a big subnet but, for example, it may be useful for modelling an ISP network, or an ISP's gateway may be in a "/56" subnet that's outside a 64-bit range used by hosts.

**- How I did it**

- Don't increment "unselected" in Bitmap when clearing a 0
  - Bug fix in the bitmap.Bitmap code
- Add addrset.AddrSet to track a set of IP addresses
  - A new internal package to track IP addresses.
  - It owns a `bitmap.Bitmap` for each range of up to 63-bits that has a bit set, within the subnet it's asked to represent.
- Use addrset.AddrSet instead of bitmap.Bitmap in IPAM

**- How to verify it**

New unit tests for `AddrSet`, plus existing tests (including the two "XFAIL" tests that now pass).

**- Description for the changelog**
```markdown changelog
- IPAM now handles subnets bigger than "/64".
```

robmry committed:
Don't increment "unselected" in Bitmap when clearing a 0

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Add addrset.AddrSet to track a set of IP addresses

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Use addrset.AddrSet instead of bitmap.Bitmap in IPAM

The default IPAM allocator is unable to represent subnets larger than
64-bits (subnets with a smaller prefix), because it uses a Bitmap
that's limited to 64-bits.

When it's used to represent a 64-bit subnet, the top address can't
be allocated (because bitmap.Bitmap is initialised with the number
of bits it needs to represent in a uint64, so it's one short).

The rest of the daemon doesn't know about these limitations, so
strange things happen when a large IPv6 subnet is used.

No errors are reported, addresses/subnets are just set up incorrectly.
The IPAM code calculates offsets into the bitmap itself, details it
shouldn't need to understand and, because it's working on offsets
into a range it doesn't always notice when it's asked to set a bit
outside the range.

It's unusual to need a big subnet but, for example, it may be useful
for modelling an ISP network, or an ISP's gateway may be in a "/56"
subnet that's outside a 64-bit range used by hosts.

So, use addrset.AddrSet instead of bitmap.Bitmap.

Signed-off-by: Rob Murray <rob.murray@docker.com>

thaJeztah commented on libnetwork/bitmap/sequence.go:
```diff
@@ -481,14 +484,16 @@ func findSequence(head *sequence, bytePos uint64) (*sequence, *sequence, uint64,
 // C) block is in the middle of current: [prev seq] [curr pre] [new] [curr post] [next seq]
-func pushReservation(bytePos, bitPos uint64, head *sequence, release bool) *sequence {
+//
+// The bool return is true if the bit value was changed.
+func pushReservation(bytePos, bitPos uint64, head *sequence, release bool) (*sequence, bool) {
```
This would be a case where possibly a named output variable could be OK;


```suggestion
// The bool return is true if the bit value was changed.
func pushReservation(bytePos, bitPos uint64, head *sequence, release bool) (_ *sequence, changed bool) {
```

or

```suggestion
// The bool return is true if the bit value was changed.
func pushReservation(bytePos, bitPos uint64, head *sequence, release bool) (newHead *sequence, changed bool) {
```

robmry replied:
Done.

thaJeztah replied:
Thanks!

thaJeztah commented on libnetwork/bitmap/sequence_test.go:
```diff
@@ -513,7 +513,7 @@ func TestPushReservation(t *testing.T) {
 
 	for n, i := range input {
-		mask := pushReservation(i.bytePos, i.bitPos, i.mask, false)
+		mask, _ := pushReservation(i.bytePos, i.bitPos, i.mask, false)
 		if !mask.equal(i.newMask) {
```
If we want to verify the new output var, I guess this could have something like;

```go
mask, changed := pushReservation(i.bytePos, i.bitPos, i.mask, false)
if mask.equal(i.mask) != changed {
....
```

Or something along those lines

robmry replied:
Ok, done. (It's also tested by the new `AddrSet` unit tests clearing unset bits.)

thaJeztah replied:
Ah, good call; thanks for updating, it was only a minor nit

thaJeztah commented on libnetwork/ipams/defaultipam/allocator_test.go:
```diff
@@ -948,11 +944,7 @@ func assertGetAddress(t *testing.T, subnet string) {
-	numAddresses := 1 << uint(zeroes)
-
-	bm := bitmap.New(uint64(numAddresses))
+	bm := addrset.New(sub)
```
Curious; can the new package fully replace (direct uses of) the  `bitmap` package? After this PR it's imported in 2 locations, so I was wondering if those would also switch over?

```bash
git grep '"github.com/docker/docker/libnetwork/bitmap"'
libnetwork/drivers/overlay/ovmanager/ovmanager.go:      "github.com/docker/docker/libnetwork/bitmap"
libnetwork/internal/addrset/addrset.go: "github.com/docker/docker/libnetwork/bitmap"
```

robmry replied:
The overlay network's bitmap is tracking vlan ids (not IP addresses).

The other one is the new `AddrSet`, which is implemented using the bitmap.

thaJeztah replied:
Ah! Gotcha! Didn't go that deep comparing 👍

thaJeztah commented on libnetwork/internal/addrset/addrset.go:
```diff
@@ -0,0 +1,198 @@
+)
+
+var (
+	// ErrBadAddress is returned when a requested address is not in the pool
+	ErrBadAddress = errors.New("address not in pool")
```
Looks like this one is not used outside of the package, so could be un-exported, or should the other check in `getAddress` map this to some  `ipamapi` type? The `ipamapi` errors look to implement `errdefs` types, whereas this one doesn't 🤔

robmry replied:
IPAM had a tendency to ask for addresses that weren't in the subnet it thought it was dealing with, and not notice the problem ... initially, I thought it might be useful for it to spot that happening, but didn't use it in the end. So, it's gone now.

thaJeztah commented on libnetwork/internal/addrset/addrset.go:
```diff
@@ -0,0 +1,198 @@
+var (
+	// ErrBadAddress is returned when a requested address is not in the pool
+	ErrBadAddress = errors.New("address not in pool")
+	// ErrNotAvailable is returned when no more addresses are available to set
+	ErrNotAvailable = errors.New("address not available")
+	// ErrAllocated is returned when the specific address requested is already allocated
+	ErrAllocated = errors.New("address already allocated")
```
I guess I was wondering if we have some option to alias some of the types so that we don't have to introduce yet-another set of errors 😅, e.g.


```suggestion
	// ErrNotAvailable is returned when no more addresses are available to set
	ErrNotAvailable = ipamapi.ErrNoAvailableIPs
	// ErrAllocated is returned when the specific address requested is already allocated
	ErrAllocated = ipamapi.ErrIPAlreadyAllocated
```

or

```suggestion
	// ErrNotAvailable is returned when no more addresses are available to set
	ErrNotAvailable = bitmap.ErrBitAllocated
	// ErrAllocated is returned when the specific address requested is already allocated
	ErrAllocated = bitmap.ErrNoBitAvailable
```

(but the bitmap errors don't implement `errdefs` types)

robmry replied:
The bitmap error texts talk about "bits" being available, but this new package is dealing with a set of addresses. So, bitmap's error messages didn't really make sense - and maybe we'll want to represent the address set differently at some point.

It didn't seem right for `AddrSet` to return IPAM errors, even though it's only used by `defaultipam` at the moment. But if you think it should, I can do that? (Could make package `addrset` internal to `defaultipam` so that it can't be used by anything else?)

thaJeztah replied:
Yeah, it was mostly wishful thinking on this one; libnetwork had a bit of a tendency to define errors for everything, even for those not actually checked for; in some cases that led to tighter coupling than needed.

So I was looking if it would be possible somehow to get closer to our errdefs definitions for matching specific conditions, but otherwise keep things generic.

This is all internal, so not a huge deal (doesn't expand public APIs / contracts.

corhere commented on libnetwork/internal/addrset/addrset.go:
```diff
@@ -0,0 +1,195 @@
+}
+
+// AddAny adds an arbitrary address to the set, and returns that address. Or, if
+// no addresses are available, it returns a wrapped [ErrNotAvailable].
```
Document the limitations of this function. An ErrNotAvailable returned from this function does not necessarily mean that all host IDs in the prefix have been allocated.

robmry replied:
As above.

corhere commented on libnetwork/internal/addrset/addrset.go:
```diff
@@ -0,0 +1,195 @@
+// If ipr is not fully contained within the set's range, it returns an error.
+//
+// When serial=true, the set is scanned starting from the address following
+// the address most recently set by [AddrSet.SetAny] or [AddrSet.SetAnyInRange].
```
```suggestion
// the address most recently set by [AddrSet.AddAny] or [AddrSet.AddAnyInRange].
```

robmry replied:
Fixed.

robmry committed:
Clarify code/comments in defaultipam.newPoolData

The first address in an IPv6 range was reserved, but that wasn't
clear from comments or the code.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented on libnetwork/internal/addrset/addrset.go:
```diff
@@ -0,0 +1,195 @@
+// Add adds address addr to the set. If addr is already in the set, it returns a
+// wrapped [ErrAllocated]. If addr is not in the set's address range, it returns
+// an error.
+func (as AddrSet) Add(addr netip.Addr) error {
```
Changed to pointers.

robmry commented on libnetwork/internal/addrset/addrset.go:
```diff
@@ -0,0 +1,195 @@
+	// If the pool has maxBitsPerBitmap or fewer, the key for the only bitmap is the pool's address.
+	// If there's more than one bitmap, they all have maxBitsPerBitmap. So, assume there's a free
+	// address in the bitmap keyed on the pool's address (no need to search other bitmaps, or work
+	// out if more bitmaps could be created).
```
The "unsound" assumption would be that anything in libnetwork is capable of tracking, or needs to track, 2^63 different IP addresses (9,223,372,036,854,775,808 addresses). Let alone more than that.

But, for the overly ambitious or concerned, I've reworded and included it in the godoc comment.

robmry commented:
I've added an extra commit, just to clarify the code that reserves network/broadcast addresses.

thaJeztah reviewed (approved):
LGTM

akerouanton commented on libnetwork/internal/addrset/addrset_test.go:
```diff
@@ -0,0 +1,215 @@
+	// Add an address that's already present. Expect an error.
+	err = as.Add(netip.MustParseAddr("10.20.255.255"))
+	assert.Check(t, is.ErrorIs(err, ErrAllocated))
+	assert.Check(t, is.ErrorIs(err, ErrAllocated))
```
```suggestion
```

robmry replied:
Oops, thank you - fixed.

akerouanton reviewed (approved):
Small nit, otherwise LGTM
